exonum/src/runtime/mod.rs

```diff
@@ -238,62 +293,49 @@ pub trait Runtime: Send + Debug + 'static {
     /// # Policy on Panics
     ///
     /// Do not process. Panic will be processed by the method caller.
-    ///
     fn execute(
         &self,
-        context: &ExecutionContext,
+        context: ExecutionContext,
         call_info: &CallInfo,
         arguments: &[u8],
     ) -> Result<(), ExecutionError>;
 
     /// Gets the state hashes of the every available service.
     fn state_hashes(&self, snapshot: &dyn Snapshot) -> StateHashAggregator;
 
-    /// Calls `before_commit` for all the services stored in the runtime.
-    ///
-    /// # Notes for Runtime Developers.
-    ///
-    /// * The order of services during invocation of this method must be the same for each node.
-    /// In other words, the order of the runtime services must be the same for each node.
-    ///
-    /// # Policy on Panics
+    /// Calls `before_commit` for a service stored in the runtime.
     ///
-    /// * Catch each kind of panics except for `FatalError` and write
-    /// them into the log.
-    /// * If panic occurs, the runtime rolls back the changes in the fork.
-    fn before_commit(&self, dispatcher: &DispatcherRef, fork: &mut Fork);
-
-    /// Calls `after_commit` for all the services stored in the runtime.
+    /// `before_commit` is called for all services active at the beginning of the block
+    /// (i.e., services instantiated within the block do **not** receive a call) exactly
+    /// once for each block.
     ///
-    /// # Policy on Panics
+    /// # Guarantees
     ///
-    /// * Catch each kind of panics except for `FatalError` and write
-    /// them into the log.
-    fn after_commit(
+    /// - Each `before_commit` call is isolated with a separate checkpoint. A call that returns
+    ///   an error will be rolled back.
+    /// - Ordering of calls among service instances is not specified, but is guaranteed
+    ///   to be the same for all nodes.
+    fn before_commit(
```

Comment:
I think that `on_resume` and `before_commit`, `after_commit` sounds a little bit inconsistent together.